src/Microsoft.ML.Data/Evaluators/Metrics/ConfusionMatrix.cs

```diff
@@ -0,0 +1,122 @@
+        /// </summary>
+        internal IReadOnlyList<ReadOnlyMemory<char>> PredictedClassesIndicators;
+
+        internal readonly bool IsSampled;
+        internal readonly bool IsBinary;
+
+        private readonly IHost _host;
+
+        /// <summary>
+        /// The confusion matrix as a structured type, built from the counts of the confusion table idv.
+        /// </summary>
+        /// <param name="host">The IHost instance. </param>
+        /// <param name="precision">The values of precision per class.</param>
+        /// <param name="recall">The vales of recall per class.</param>
+        /// <param name="confusionTableCounts">The counts of the confusion table. The actual classes values are in the rows of the 2D array,
+        /// and the counts of the predicted classes are in the columns.</param>
+        /// <param name="labelNames">The predicted classes names, or the indexes of the classes, if the names are missing.</param>
+        /// <param name="isSampled">Whether the classes are sampled.</param>
+        /// <param name="isBinary">Whether the confusion table is the result of a binary classification. </param>
+        /// <param name="classIndicators">The Annotations of the Count column, in the confusionTable idv.</param>
+        internal ConfusionMatrix(IHost host, double[] precision, double[] recall, double[][] confusionTableCounts,
+            List<ReadOnlyMemory<char>> labelNames, bool isSampled, bool isBinary, DataViewSchema.Annotations classIndicators)
+        {
+            _host = host;
+
+            _host.AssertNonEmpty(precision);
+            _host.AssertNonEmpty(recall);
+            _host.AssertNonEmpty(confusionTableCounts);
+            _host.AssertNonEmpty(labelNames);
+            _host.AssertNonEmpty(precision);
+
+            _host.Assert(precision.Length == confusionTableCounts.Length);
+            _host.Assert(recall.Length == confusionTableCounts.Length);
+            _host.Assert(labelNames.Count == confusionTableCounts.Length);
+
+            PerClassPrecision = precision.ToImmutableArray();
+            PerClassRecall = recall.ToImmutableArray();
+            IsSampled = isSampled;
+            IsBinary = isBinary;
+            PredictedClassesIndicators = labelNames.AsReadOnly();
+
+            NumberOfPredictedClasses = confusionTableCounts.Length;
+            List<ImmutableArray<double>> counts = new List<ImmutableArray<double>>(NumberOfPredictedClasses);
+
+            for (int i = 0; i < NumberOfPredictedClasses; i++)
+                counts.Add(ImmutableArray.Create(confusionTableCounts[i]));
+
+            Counts = counts.ToImmutableArray();
+            ClassIndicators = classIndicators;
+        }
+
+        /// <summary>
+        /// Returns a human readable representation of the confusion table.
+        /// </summary>
+        /// <returns></returns>
+        public string GetFormattedConfusionTable() => MetricWriter.GetConfusionTableAsString(this, false);
```

Comment:
>MetricWriter.GetConfusionTableAsString [](start = 54, length = 38)

Can we cache it?
Or just calculate it constructor?
I'm not sure in which case this would produce you different results in different times of call.

Reply:
caching it and calculating it when first asked for.

---
In reply to: [275469566](https://github.com/dotnet/machinelearning/pull/3250#discussion_r275469566) [](ancestors = 275469566)